src/main/java/org/truffleruby/parser/PatternMatchingTranslator.java

```diff
@@ -0,0 +1,439 @@
+import org.truffleruby.RubyLanguage;
+import org.truffleruby.core.array.ArrayIndexNodes;
+import org.truffleruby.core.array.ArrayLiteralNode;
+import org.truffleruby.core.array.ArrayPatternLengthCheckNode;
+import org.truffleruby.core.array.ArraySliceNodeGen;
+import org.truffleruby.language.RubyNode;
+import org.truffleruby.language.SourceIndexLength;
+import org.truffleruby.language.arguments.EmptyArgumentsDescriptor;
+import org.truffleruby.language.control.AndNode;
+import org.truffleruby.language.control.ExecuteAndReturnTrueNode;
+import org.truffleruby.language.control.NotNode;
+import org.truffleruby.language.control.RaiseException;
+import org.truffleruby.language.dispatch.RubyCallNodeParameters;
+import org.truffleruby.language.literal.BooleanLiteralNode;
+import org.truffleruby.language.literal.NilLiteralNode;
+import org.truffleruby.language.literal.TruffleInternalModuleLiteralNode;
+import org.truffleruby.language.locals.ReadLocalNode;
+import org.truffleruby.language.locals.WriteLocalNode;
+import org.truffleruby.parser.ast.ArrayParseNode;
+import org.truffleruby.parser.ast.ArrayPatternParseNode;
+import org.truffleruby.parser.ast.CallParseNode;
+import org.truffleruby.parser.ast.ConstParseNode;
+import org.truffleruby.parser.ast.DAsgnParseNode;
+import org.truffleruby.parser.ast.DStrParseNode;
+import org.truffleruby.parser.ast.DotParseNode;
+import org.truffleruby.parser.ast.FalseParseNode;
+import org.truffleruby.parser.ast.FindPatternParseNode;
+import org.truffleruby.parser.ast.FixnumParseNode;
+import org.truffleruby.parser.ast.HashPatternParseNode;
+import org.truffleruby.parser.ast.IfParseNode;
+import org.truffleruby.parser.ast.LambdaParseNode;
+import org.truffleruby.parser.ast.ListParseNode;
+import org.truffleruby.parser.ast.LocalAsgnParseNode;
+import org.truffleruby.parser.ast.NilParseNode;
+import org.truffleruby.parser.ast.ParseNode;
+
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.NodeUtil;
+import com.oracle.truffle.api.source.Source;
+import org.truffleruby.parser.ast.RegexpParseNode;
+import org.truffleruby.parser.ast.StarParseNode;
+import org.truffleruby.parser.ast.StrParseNode;
+import org.truffleruby.parser.ast.TrueParseNode;
+
+import java.util.Arrays;
+
+public class PatternMatchingTranslator extends BaseTranslator {
+
+    ParseNode data;
+    ListParseNode cases;
+    TranslatorEnvironment environment;
+    BodyTranslator bodyTranslator;
+
+    RubyNode currentValueToMatch;
+
+    public PatternMatchingTranslator(
+            RubyLanguage language,
+            Source source,
+            ParserContext parserContext,
+            Node currentNode,
+            ParseNode data,
+            ListParseNode cases,
+            TranslatorEnvironment environment,
+            BodyTranslator bodyTranslator) {
+        super(language, source, parserContext, currentNode, environment);
+        this.data = data; // data to match on
+        this.cases = cases; // cases to check.
```

Comment:
These comments would be better on the fields above, as javadoc comments (`/** foo */`)

Reply:
converted to javadoc comments